xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,288 @@
+          }
+        }
+
+        @Override
+        public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+          if (clusterWatcher == null) {
+            clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+            xdsClient.watchClusterData(cdsConfig.name, clusterWatcher);
+            if (oldCdsConfig != null) {
+              xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+            }
+            CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+          }
+          // Else handleResolvedAddresses() already called, so no-op here because the config is
+          // fixed in this balancer.
+        }
+      };
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    // EDS balancer for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadBalancer edsBalancer;
+
+    // LoadStatsStore for the cluster.
+    // Becomes non-null once handleResolvedAddresses() successfully.
+    // Assigned at most once.
+    @Nullable
+    LoadStatsStore loadStatsStore;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate newUpdate) {
+      channelLogger.log(ChannelLogLevel.DEBUG, "Received a CDS update: '%s'",  newUpdate);
+      checkArgument(
+          newUpdate.getLbPolicy().equals("round_robin"),
+          "The load balancing policy in ClusterUpdate '%s' is not supported", newUpdate);
+
+      final XdsConfig edsConfig = new XdsConfig(
+          /* balancerName = */ null,
+          new LbConfig(newUpdate.getLbPolicy(), ImmutableMap.<String, Object>of()),
+          /* fallbackPolicy = */ null,
+          /* edsServiceName = */ newUpdate.getEdsServiceName(),
+          /* lrsServerName = */ newUpdate.getLrsServerName());
+
+      if (loadStatsStore == null) {
+        loadStatsStore = new LoadStatsStoreImpl();
+      }
+      if (edsBalancer == null) {
+        edsBalancer = lbRegistry.getProvider(XDS_POLICY_NAME).newLoadBalancer(helper);
+      }
+      edsBalancer.handleResolvedAddresses(
+          resolvedAddresses.toBuilder()
+              .setAttributes(
+                  resolvedAddresses.getAttributes().toBuilder()
+                      .discard(ATTR_LOAD_BALANCING_CONFIG)
+                      .set(XdsAttributes.LOAD_STATS_STORE_REF, loadStatsStore)
+                      .build())
+              .setLoadBalancingPolicyConfig(edsConfig)
+          .build());
+    }
+
+    @Override
+    public void onError(Status error) {
+      channelLogger.log(ChannelLogLevel.ERROR, "Received a CDS error: '%s'",  error);
+
+      // Go into TRANSIENT_FAILURE if we have not yet created the child
```

Comment:
child balancer? Do you mean downstream balancer?

Reply:
The child policy is EDS policy. What do you mean by downstream balancer? We didn't use that terminology.

Reply:
We can use child balancer and child policy interchangeably here, because once you create an EDS policy, you create an edsConfig and pass it to handleResolvedAddresses immediately.